src/components/views/auth/InteractiveAuthEntryComponents.js

```diff
@@ -564,6 +600,88 @@ export const MsisdnAuthEntry = createReactClass({
     },
 });
 
+export class SSOAuthEntry extends React.Component {
+    static propTypes = {
+        matrixClient: PropTypes.object.isRequired,
+        authSessionId: PropTypes.string.isRequired,
+        loginType: PropTypes.string.isRequired,
+        submitAuthDict: PropTypes.func.isRequired,
+        errorText: PropTypes.string,
+        onPhaseChange: PropTypes.func.isRequired,
+        continueText: PropTypes.string,
+        continueKind: PropTypes.string,
+        onCancel: PropTypes.func,
+    };
+
+    static LOGIN_TYPE = "m.login.sso";
+    static UNSTABLE_LOGIN_TYPE = "org.matrix.login.sso";
+
+    static PHASE_PREAUTH = 1; // button to start SSO
+    static PHASE_POSTAUTH = 2; // button to confirm SSO completed
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            // We actually send the user through fallback auth so we don't have to
+            // deal with a redirect back to us, losing application context.
```

Comment:
I'm not sure I understand why using a fallback URL prevents losing context - isn't the thing that prevents us from losing context just having the 'confirm' button that the user has to come back & click?

Reply:
see https://github.com/matrix-org/matrix-react-sdk/pull/4292#discussion_r400956782